### DISCUSSION TITLE
Question: Is it ok that running queries in asyncio tasks concurrently (via `gather`) result in multiple short lived connections in `AsyncAdaptedQueuePool`?

### DISCUSSION THREAD

**Discussion by somnam** [sqlalchemy/sqlalchemy#10122] [Usage Questions] [answered]
I'm running concurrent queries to Postgresql DB using `psycopg` version 3 with the `async` mode enabled via `postgresql+psycopg_async`. 
After running a set of queries I've noticed that the `pid` in `pg_stat_activity` is increasing by a noticeable factor.
The code I run is very similar to the following:
```python
url = "postgresql+psycopg_async://user:pass@database:5432/postgres?application_name=async-sqla-session"

engine = create_async_engine(url)

session_maker = async_sessionmaker(engine)

@contextlib.asynccontextmanager
async def session_scope() -> AsyncGenerator:
    async with session_maker() as session, session.begin():
        yield session

async def example_task() -> None:
    async with session_scope() as session:
        await session.scalar(text("SELECT pg_sleep(0.1);"))  # simulate a running query

async def main() -> None:
    await asyncio.gather(*[example_task() for _ in range(128)])

    stmt = text("SELECT pid FROM pg_stat_activity WHERE application_name = 'async-sqla-session';")
    async with session_scope() as session:
        print(f"Last conn pid: {max(await session.scalars(stmt))}")

asyncio.run(main())
```

The last connection pid is printed at the end of the script. A few runs result in this value increasing quite noticeably:
```shell
$ python src/async_sqla_session.py
Last conn pid: 129
$ python src/async_sqla_session.py
Last conn pid: 144
$ python src/async_sqla_session.py
Last conn pid: 158
$ python src/async_sqla_session.py
Last conn pid: 191
$ python src/async_sqla_session.py
Last conn pid: 252
$ python src/async_sqla_session.py
Last conn pid: 309
$ python src/async_sqla_session.py
Last conn pid: 357
```
I was assuming that each script run should result in a lower number of created connections (~ 15?) than as seen (~ 50).
Maybe the connection pool can't keep up with the amount of new requests that happen when all tasks are gathered at the same time and awaited concurrently so it needs to close / spawn new connections constantly?
Is this normal behavior for this type of operations when using an async engine / driver?

**Comment by CaselIT** [comment]
Hi,

You can enable the pool logging in create engine to look at what the pool is doing 
https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.echo_pool

I don't think looking at the pid is a accurate way of judging the number of connections, since that should be handled by the os

**Comment by zzzeek** [comment, marked as answer]
what you are seeing is likely pool overflow; the pool only keeps 5 connections by default, the other 10 are opened and closed on each use, so that would be why the number keeps increasing (somewhat randomly, depending on how the gather works out, I guess).

Try setting pool_size=15 and max_overflow=0.  then the pid thing should be fixed at 15 or less.

**Reply by CaselIT** [reply]
forgot about the overflow

**Reply by somnam** [reply]
Thank you, setting `pool_size=15 and max_overflow=0` does keep a fixed pool of connections.
Would you recommend one of these approaches over the other (fixed pool vs small pool with overflow) in an asyncio app like Fastapi?

**Reply by zzzeek** [reply]
for a web application I'd normally combine pool size and overflow as the model there is db-connection-per-web-request , so if the  application is not getting many requests, using overflow will keep the number of waiting connections on the server lower.

OTOH if I'm throwing many connection-per-task into gather quite a lot in the way that's illustrated here, i have the option to send batches to gather that are less than the pool size so that connections are always re-used.     

if one is combining any number of FastAPI requests, *and* a gather operation within the scope of a request, scaling that is trickier, I'd mostly want to be careful about how many concurrent requests FastAPI is allowing combined with what the largest batch a request might send at a time for gather.

finally, with gather and DB work, as we recently figured out at #9312, I would at the very least use gather **only** with `return_exceptions=True` and preferably use `TaskGroup` and not gather.   otherwise if any of your tasks raise an exception, your application will leak connections which will lead to warnings and possibly even deadlocks.

**Comment by somnam** [comment]
Thanks a lot for the detailed answer!
